src/scope.rs

```diff
@@ -1,299 +1,530 @@
+//! it's wrapping, and `'s` is the lifetime of the scope.
+//!
+//! Effectively you have `&'b PinnedRef<'s, T>`.
+//!
+//! The lifetime for handles bound to the scope is the lifetime of the scope that it was created from.
+//! So in `ContextScope<'b, 's>`, `'b` is the lifetime of the borrow of the inner scope, and `'s` is the lifetime of the inner scope (and therefore the handles).
+use crate::{
+  Context, Data, DataError, Function, FunctionCallbackInfo, Isolate, Local,
+  Message, Object, OwnedIsolate, PromiseRejectMessage, PropertyCallbackInfo,
+  SealedLocal, Value, fast_api::FastApiCallbackOptions, isolate::RealIsolate,
+  support::assert_layout_subset,
+};
+use std::{
+  any::type_name,
+  cell::Cell,
+  marker::{PhantomData, PhantomPinned},
+  mem::ManuallyDrop,
+  ops::{Deref, DerefMut},
+  pin::Pin,
+  ptr::NonNull,
+};
+pub(crate) mod raw;
+
+pub type PinScope<'s, 'i, C = Context> = PinnedRef<'s, HandleScope<'i, C>>;
+pub type PinCallbackScope<'s, 'i, C = Context> =
+  PinnedRef<'s, CallbackScope<'i, C>>;
+
+/// Storage for a scope.
+///
+/// Tracks the initialization state of the scope, and holds the scope itself.
+#[repr(C)]
+pub struct ScopeStorage<T: ScopeInit> {
+  inited: bool,
+  scope: ManuallyDrop<T>,
+  _pinned: PhantomPinned,
 }
 
-impl<'s, P: param::NewContextScope<'s>> ContextScope<'s, P> {
-  #[allow(clippy::new_ret_no_self)]
-  pub fn new(param: &'s mut P, context: Local<Context>) -> P::NewScope {
-    let scope_data = param.get_scope_data_mut();
-    if scope_data.get_isolate_ptr()
-      != unsafe { raw::v8__Context__GetIsolate(&*context) }
-    {
-      panic!(
-        "{} and Context do not belong to the same Isolate",
-        type_name::<P>()
-      )
+impl<T: ScopeInit> ScopeStorage<T> {
+  pub(crate) fn projected(self: Pin<&mut Self>) -> Pin<&mut T> {
+    // SAFETY: we are just projecting to a field, so the scope remains pinned
+    unsafe {
+      let self_mut = self.get_unchecked_mut();
+      Pin::new_unchecked(&mut self_mut.scope)
+    }
+  }
+
+  pub fn new(scope: T) -> Self {
+    Self {
+      inited: false,
+      scope: ManuallyDrop::new(scope),
+      _pinned: PhantomPinned,
+    }
+  }
+
+  pub fn init(mut self: Pin<&mut Self>) -> PinnedRef<'_, T> {
+    if self.inited {
+      // free old, going to reuse this storage
+      unsafe {
+        let self_mut = self.as_mut().get_unchecked_mut();
+        self_mut.drop_inner();
+        self_mut.inited = false;
+      }
+    }
+
+    // hold onto a pointer so we can set this after initialization. we can't use a normal
+    // mutable reference because the borrow checker will see overlapping borrows. this is
+    // safe, however, because we lose our mutable reference to the storage in `init_stack`
+    // as it gets projected to the inner type
+    let inited_ptr =
+      unsafe { &raw mut self.as_mut().get_unchecked_mut().inited };
+    let ret = T::init_stack(self);
+    unsafe { inited_ptr.write(true) };
```

Comment:
Have you tried making `initited` a `Cell<bool>`? so this won't need unsafe code:

```rust
let ret = T::init_stack(self);
self.as_ref().inited.set(true);
```